pkg/proxy/nftables/proxier.go

```diff
@@ -975,6 +998,89 @@ func isAffinitySetName(set string) bool {
 	return strings.HasPrefix(set, servicePortEndpointAffinityNamePrefix)
 }
 
+type containerArray interface {
+	[0]string | [1]string | [2]string | [3]string
+}
+
+// nftContainer is an internal representation of map or set.
+type nftContainer[T1, T2 containerArray] struct {
+	elements      map[T1]T2
+	leftoverKeys  sets.Set[T1]
+	containerType string
+	containerName string
+}
+
+// newNftContainer creates an empty nftContainer and flushes the corresponding nftables container.
+func newNftContainer[T1, T2 containerArray](containerType, containerName string, tx *knftables.Transaction) *nftContainer[T1, T2] {
+	c := &nftContainer[T1, T2]{
+		elements:      make(map[T1]T2),
+		leftoverKeys:  sets.New[T1](),
+		containerType: containerType,
+		containerName: containerName,
+	}
+	if containerType == "set" {
```

Comment:
if we add independent types for sets and maps instead of trying to generalize?

Reply:
can do that, the only disadvantage is the need to duplicate all functions and most of the logic for maps and sets.

Reply:
it is an ignorant question 😄 , if both maps and sets have the same operations 🤔

Reply:
hmm, I personally think this should be an interface
```
type nftElementStorage interface {
  ensureElem()
  resetLeftoverKeys()
  cleanupLeftoverKeys()
  reset(). <--- https://github.com/kubernetes/kubernetes/pull/126013/commits/cb279fbf4f6697c6edf07b0432e8609159417868#r1682935666
  update() <-- https://github.com/kubernetes/kubernetes/pull/126013/commits/cb279fbf4f6697c6edf07b0432e8609159417868#r1682935666
}
```

and then you define the `sets` and `maps` types that implement that interface

Reply:
With the single implementation I am trying to avoid code duplication, as set logic is a simplified copy of a map logic. I prefer having a couple of `if`s and share the code, but lmk if you still think having 2 copies makes more sense.
I did add a `reset()` method to the `nftElementStorage`.

Reply:
If maps and sets are different is likely we evolve one or the other inside the same object forcing us to refactor later.

Is not that there is a lot of code that gets duplicated, and this allows to proper unit test each object

Reply:
That is exactly the point, set **is** just a map without values, so I don't think we will ever evolve them in different directions. If you check current usage of `s.containerType`, it is only used to set the right knftables fields, but not to change the logic.
For example, based on Dan's comment, I have moved `resetLeftoverKeys` call to other functions, and this change is exactly the same for set and map.
Also about unit tests: currently every unit test using map automatically tests the logic for sets, but if we split them, we will need 2 times more tests to cover both sets and maps separately.
Maybe I am missing something, but I don't see any advantages of splitting it for now. We can still split it in the future if needed though.